### PR TITLE
[TK-2028] 🚀 Add a new controller `AgentPool`

### PR DESCRIPTION
**Pull request by arybolovlev**
### Description

This PR adds a new controller that allows manages Terraform Cloud Agent Pools(more information [here](https://developer.hashicorp.com/terraform/cloud-docs/agents/agent-pools)).

The controller creates an agent pool and serves as a single source of truth to manage it. It means that all manual changes will be reverted to the state described in the CR spec.

The controller stores created tokens in related Kubernetes Secrets, each CR created its own Secret within the same namespace as the CR. Once CR is removed, the related Secret will be removed as well.

#### CRD

```yaml
apiVersion: app.terraform.io/v1alpha2
kind: AgentPool
metadata:
  name: <STRING>
  namespace: <STRING>
spec:
  name: <STRING>
  organization: <STRING>
  token:
    secretKeyRef:
      name: <STRING>
      key: <STRING>
  agentTokens: <LIST>
    - name: <STRING>
status:
  observedGeneration: <NUMBER>
  agentPoolID: <STRING>
  agentTokens: <LIST>
    - name: <STRING>
      id: <STRING>
      createdAt: <NUMBER>
      lastUsedAt: <NUMBER>
```

### Usage Example
Create a new Agent Pool `production`:
```yaml
apiVersion: app.terraform.io/v1alpha2
kind: AgentPool
metadata:
  name: production
spec:
  name: production
  organization: kubernetes-operator
  token:
    secretKeyRef:
      name: tfe-operator
      key: token
  agentTokens:
    - name: eu-nl-01
    - name: eu-de-01
    - name: eu-de-02
    - name: us-ca-01
    - name: us-ca-02
```

Once it's created, Agent Tokens can be found in Kubernetes Secret `production-agent-pool`:
```yaml
apiVersion: v1
data:
apiVersion: v1
data:
  eu-de-01: V3V6T
  eu-de-02: Z0FGN
  eu-nl-01: TXJhY
  us-ca-01: ekRBa
  us-ca-02: ZHdEe
kind: Secret
metadata:
  creationTimestamp: "2022-12-02T12:47:12Z"
  labels:
    agentPoolID: apool-2qHqZ1VUkf5Qr4aq
  name: production-agent-pool
  namespace: hashistack
  ownerReferences:
  - apiVersion: app.terraform.io/v1alpha2
    blockOwnerDeletion: true
    controller: true
    kind: AgentPool
    name: production
    uid: 33a2c085-7efe-412d-988a-a1c151f9d7be
  resourceVersion: "712592"
  uid: 49784132-c9e0-46d3-a9fd-d22be5459643
type: Opaque
```

### Release Note
Release note for [CHANGELOG](https://github.com/hashicorp/terraform-cloud-operator/blob/main/CHANGELOG.md):
```release-note
[NEW FEATURE] Add a new controller `AgentPool` to manage TFC Agent Pools.
[ENHANCEMENT] Bump dependencies to the latest available versions.
[ENHANCEMENT] Add the ability to run tests on the real cluster if environment variable `USE_EXISTING_CLUSTER` is set to `true`: `export USE_EXISTING_CLUSTER=true`.
```

### References
N/A

### Community Note
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request.
* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -104,7 +104,7 @@ vet: ## Run go vet against code.
 
 .PHONY: test
 test: manifests generate fmt vet envtest ## Run tests.
-	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path)" go test -timeout 15m -v ./controllers -coverprofile cover.out
+	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path)" go test -timeout 30m -v ./controllers -coverprofile cover.out
```
Comment (arybolovlev) [inline Makefile:107]: I make the timeout 30 minutes to make sure all tests have to complete. That can be changed later when tests will be executed in parallel.

File: Makefile [hunk 2]
```diff
@@ -160,8 +160,8 @@ CONTROLLER_GEN ?= $(LOCALBIN)/controller-gen
 ENVTEST ?= $(LOCALBIN)/setup-envtest
 
 ## Tool Versions
-KUSTOMIZE_VERSION ?= v4.5.5
-CONTROLLER_TOOLS_VERSION ?= v0.9.0
+KUSTOMIZE_VERSION ?= v4.5.7
+CONTROLLER_TOOLS_VERSION ?= v0.10.0
```
Comment (arybolovlev) [inline Makefile:164]: This is not related to the `AgentPool` controller.

**Review by jrhouston — APPROVED** [review]
